### PR TITLE
support PHPUnit 10, PHP 8.1+, cleanup

### PR DESCRIPTION
**Pull request by codemasher**
So that's the cleanup PR mentioned in #47. It removes the random abstract classes and traits, updates to some modern PHPUnit features (e.g. data provider calls via attributes) and generally some "modern" PHP features such as typed properties.

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -13,10 +13,14 @@
         {
             "name": "PHP-FIG",
             "homepage": "https://www.php-fig.org/"
+        },
+        {
+            "name": "Contributors",
+            "homepage":"https://github.com/http-interop/http-factory-tests/graphs/contributors"
```
Comment (shadowhand) [inline composer.json]: 👍🏼

File: test/RequestFactoryTest.php [hunk 2]
```diff
@@ -2,34 +2,62 @@
...
+		if(!defined('REQUEST_FACTORY') || !class_exists(REQUEST_FACTORY)){
+			static::markTestSkipped('REQUEST_FACTORY class name not provided');
+		}
+
+		if(!defined('URI_FACTORY') || !class_exists(URI_FACTORY)){
+			static::markTestSkipped('URI_FACTORY class name not provided');
+		}
+
+		$this->requestFactory = new (REQUEST_FACTORY);
+		$this->uriFactory     = new (URI_FACTORY);
```
Comment (shadowhand) [inline test/RequestFactoryTest.php]: This change is out of scope with supporting PHPUnit 10. Please revert it.

File: test/ResponseFactoryTest.php [hunk 3]
```diff
@@ -2,19 +2,42 @@
...
+	/**
+	 * @noinspection PhpUndefinedConstantInspection
+	 */
+	public function setUp():void{
+
+		if(!defined('RESPONSE_FACTORY') || !class_exists(RESPONSE_FACTORY)){
+			static::markTestSkipped('RESPONSE_FACTORY class name not provided');
+		}
+
+		$this->responseFactory = new (RESPONSE_FACTORY);
```
Comment (shadowhand) [inline test/ResponseFactoryTest.php]: This change is out of scope. Please revert it. Same goes for the rest of this PR.

The only changes should be to composer.json and anything that is necessary for PHPUnit to not give warnings about deprecations.

**Review by shadowhand — CHANGES_REQUESTED** [review]

File: test/RequestFactoryTest.php (hunk 2 above)
Comment (codemasher) [reply test/RequestFactoryTest.php]: The scope of this change is given in the OP.

Comment (shadowhand) [reply test/RequestFactoryTest.php]: > random abstract classes and traits

These are not "random", they are design decisions that have downstream impacts. Changing this package's API has consequences.

Comment (codemasher) [reply test/RequestFactoryTest.php]: Ok, so I checked the most popular dependents for this library and all except one are using it as a stand-alone test suite without any changes. The Slim framework is the only one that extends the test classes and because of the "design decision" it has to unnecessarily re-implement several methods that are only found in the finalized classes here. Since you want to increment the major version you should take the opportunity to amend bad design decisions. What i can do is to rename the classes back to "*TestCase" so that libraries that extend them won't break - so even with out changes after an upgrade, their tests will still run and the old methods are ignored.

Comment (shadowhand) [reply test/RequestFactoryTest.php]: There are two separate concerns that need to be separated:

1. #47 wasn't a complete switch to PHPUnit 10. I merged it and reduced the scope to be PHP 8.0+ and PHPUnit 10.0+ and released it as v2.0.0. Obviously that was not quite sufficient, given that PHPUnit 10 requires PHP 8.1. So one change should be _just_ finalizing the PHPUnit 10.x upgrade, with as few code changes as possible, so that we can release v2.0.1.
2. This PR shows that we could make some code/API changes that would improve the test cases, and I am absolutely for that. But such changes would be represent version 3.x, because this package follows semantic versioning. Those changes need to be a separate PR that downstream users can review and have input on.

I am happy to review and merge fixes for both (1) and (2) but they need to be separate PRs.

**Comment by codemasher** [comment]
Ok, I've changed the PR to the requested changes that should have been discussed and made in #47:

- full PHPUnit 10 support, PHP 8.1+, no changes to the original v0.x API
- call data providers via attributes
- static calls to PHPUnit assert methods
- additional `class_exist` checks when calling a class provided via PHPUnit constant
- instance classes from expression (`new (REQUEST_FACTORY)`)
- added array keys in data providers for improved PHPUnit output
- added function/constant imports

If you don't mind I'll add another commit with the changes that were supposed to be in this PR, aka, API changes.

**Review by shadowhand — APPROVED** [review]
Excellent clean up, thank you!

**Comment by shadowhand** [comment]
> If you don't mind I'll add another commit with the changes that were supposed to be in this PR, aka, API changes.

As a separate PR please.

**Comment by codemasher** [comment]
Good. Better tag it right now because [there are already people waiting for it](https://github.com/cakephp/cakephp/commit/d4b0a2a2c9692079c35c8ceb6d0403ccc1ec94cb).

**Comment by codemasher** [comment]
A small oversight on my end, i forgot a `class_exists` check over here:

https://github.com/http-interop/http-factory-tests/blob/45648c690640118a887d4c4083a580649e13c19d/test/RequestFactoryTest.php#L30
